### PR TITLE
Reintroduce Allreduce gradient

### PR DESCRIPTION
**Pull request by PhilipVinc**
This works by defining an adjoint for the `jax.lax.create_token` primitive.
Seems to work fine, though maybe @matthj could tell us if what we are doing is fine.

### CODE REVIEW

**Comment by PhilipVinc** [comment]
I'll merge for now because I urgently need this...

**Comment by dionhaefner** [comment]
I think you mean @mattjj :)

Looks OK to me, do you think we could do the same with batching?

**Comment by PhilipVinc** [comment]
Ooooups! yes indeed

Probably yes, but it requires a bit more thought as you can't have jax arrays of tokens
